uxarray/grid/geometry.py

```diff
@@ -582,3 +599,346 @@ def _insert_pt_in_latlonbox(old_box, new_pt, is_lon_periodic=True):
+                )
+                face_latlon_array[0][0] = (
+                    -np.pi / 2
+                )  # Ensure south pole is the lower latitude bound
+
+        # Adjust longitude bounds globally if the pole is centrally inside the polygon
+        if is_center_pole:
+            face_latlon_array[1] = [0.0, 2 * np.pi]
+
+    else:
+        # Normal Face
+        # Iterate through each edge_cart of a face to update the bounding box (latlonbox) with extreme latitudes and longitudes
+        for i in range(face_edges_cartesian.shape[0]):
+            edge_cart = face_edges_cartesian[i]
+            edge_lonlat = face_edges_lonlat_rad[i]
+
+            # Skip processing if the edge_cart is marked as a dummy with a fill value
+            if np.any(edge_cart == INT_FILL_VALUE):
+                continue
+
+            # Extract cartesian coordinates of the edge_cart's endpoints
+            n1_cart, n2_cart = edge_cart
+            n1_lonlat, n2_lonlat = edge_lonlat
+
+            # Convert latitudes and longitudes of the nodes to radians
+            node1_lon_rad, node1_lat_rad = n1_lonlat
+            node2_lon_rad, node2_lat_rad = n2_lonlat
+
+            # Determine if the edge_cart's extreme latitudes need to be considered using the corrected logic
+            is_GCA = (
+                is_GCA_list[i]
+                if is_GCA_list is not None
+                else not is_latlonface or n1_cart[2] != n2_cart[2]
+            )
+
+            lat_max, lat_min = (
+                (
+                    extreme_gca_latitude(np.array([n1_cart, n2_cart]), "max"),
+                    extreme_gca_latitude(np.array([n1_cart, n2_cart]), "min"),
+                )
+                if is_GCA
+                else (node1_lat_rad, node1_lat_rad)
+            )
+
+            # Insert extreme latitude points into the latlonbox if they differ from the node latitudes
+            if not np.isclose(
+                node1_lat_rad, lat_max, atol=ERROR_TOLERANCE
+            ) and not np.isclose(node2_lat_rad, lat_max, atol=ERROR_TOLERANCE):
+                # Insert the maximum latitude
+                face_latlon_array = _insert_pt_in_latlonbox(
+                    face_latlon_array, np.array([lat_max, node1_lon_rad])
+                )
+            elif not np.isclose(
+                node1_lat_rad, lat_min, atol=ERROR_TOLERANCE
+            ) and not np.isclose(node2_lat_rad, lat_min, atol=ERROR_TOLERANCE):
+                # Insert the minimum latitude
+                face_latlon_array = _insert_pt_in_latlonbox(
+                    face_latlon_array, np.array([lat_min, node1_lon_rad])
+                )
+            else:
+                # Insert the node's latitude and longitude as it matches the extreme latitudes
+                face_latlon_array = _insert_pt_in_latlonbox(
+                    face_latlon_array, np.array([node1_lat_rad, node1_lon_rad])
+                )
+
+    return face_latlon_array
```

Comment:
It would be really helpful for future optimizations to decompose this large code block into smaller, more tangible, helper functions.